### PR TITLE
IGNITE-20976: Sql. Fixed clone method for DDL/DML statements

### PR DESCRIPTION
**Pull request by lowka**
Fixes clone method of DDL/DML nodes that is the root cause of the issue.

- Fixes DML nodes provided by calcite.
- Fixes custom operators implemented in apache ignite 3.
- Fixes clone operator for `ALTER COLUMN DROP DEFAULT` to produce identical node.

https://issues.apache.org/jira/browse/IGNITE-20976

---
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/IgniteSqlCreateZone.java [hunk 1]
```diff
@@ -35,14 +35,31 @@
  * Parse tree for {@code CREATE ZONE} statement with Ignite specific features.
  */
 public class IgniteSqlCreateZone extends SqlCreate {
+
+    /** CREATE ZONE operator. */
+    protected static class Operator extends IgniteDdlOperator {
+
+        /** Constructor. */
+        protected Operator(boolean existFlag) {
+            super("ALTER ZONE", SqlKind.OTHER_DDL, existFlag);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/IgniteSqlCreateZone.java:44]: CREATE

**Review by korlov42 — APPROVED** [review]

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlAlterColumnDdlParserTest.java [hunk 2]
```diff
@@ -53,8 +46,15 @@ public class SqlAlterColumnDdlParserTest extends AbstractDdlParserTest {
      */
     @Test
     public void testNotNull() {
-        assertThat(parseAlterColumn("SET NOT NULL").notNull(), is(true));
-        assertThat(parseAlterColumn("DROP NOT NULL").notNull(), is(false));
+        IgniteSqlAlterColumn alterColumn = parseAlterColumn("ALTER TABLE t ALTER COLUMN a SET NOT NULL");
+        assertThat(alterColumn.notNull(), is(true));
+        expectUnparsed(alterColumn, "ALTER TABLE \"T\" ALTER COLUMN \"A\" SET NOT NULL");
+
+        assertThat(alterColumn.notNull(), is(true));
```
[the remark this answers is not in the document]
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlAlterColumnDdlParserTest.java]: Thanks. Fixed.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/IgniteSqlCreateZone.java (hunk 1 above)
Comment (lowka) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/IgniteSqlCreateZone.java:44]: Thanks. Fixed.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/IgniteSqlParserTest.java [hunk 3]
```diff
@@ -148,6 +150,51 @@ public void testInvalidDecimalLiteral() {
                 () -> IgniteSqlParser.parse("SELECT decimal '2a'", StatementParseResult.MODE));
     }
 
+    @ParameterizedTest
+    @MethodSource("dmlStatements")
+    public void testDmlStatementsClone(String stmt) {
+        StatementParseResult single = IgniteSqlParser.parse(stmt, StatementParseResult.MODE);
```
Comment (xtern) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/IgniteSqlParserTest.java:156]: Seems that similar test for DDL statements fails.
For example  `DROP TABLE TEST` produces the following error:

```
	at org.apache.calcite.util.Util.needToImplement(Util.java:1119)
	at org.apache.calcite.sql.SqlSyntax$7.unparse(SqlSyntax.java:129)
	at org.apache.calcite.sql.SqlOperator.unparse(SqlOperator.java:385)
	at org.apache.calcite.sql.SqlDialect.unparseCall(SqlDialect.java:466)
	at org.apache.calcite.sql.SqlCall.unparse(SqlCall.java:131)
	at org.apache.calcite.sql.SqlNode.toSqlString(SqlNode.java:156)
	at org.apache.calcite.sql.SqlNode.toString(SqlNode.java:131)
```

Is it ok?

Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/IgniteSqlParserTest.java:156]: Thanks, fixed.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlAlterColumnDdlParserTest.java [hunk 4]
```diff
@@ -96,22 +101,42 @@ public void testDefault() {
...
+
+        alterColumn = parseAlterColumn("ALTER TABLE t ALTER COLUMN c SET DATA TYPE INTEGER NULL");
+        expectDataType(alterColumn, "INTEGER", false, null);
+        expectUnparsed(alterColumn, "ALTER TABLE \"T\" ALTER COLUMN \"C\" SET DATA TYPE INTEGER NULL");
+
+        alterColumn = parseAlterColumn("ALTER TABLE t ALTER COLUMN c SET DATA TYPE INTEGER DEFAULT -1");
+        expectDataType(alterColumn, "INTEGER", null, -1);
+        expectUnparsed(alterColumn, "ALTER TABLE \"T\" ALTER COLUMN \"C\" SET DATA TYPE INTEGER DEFAULT -1");
+
+        alterColumn = parseAlterColumn("ALTER TABLE t ALTER COLUMN c SET DATA TYPE INTEGER NOT NULL DEFAULT -1");
```
Comment (xtern) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlAlterColumnDdlParserTest.java:120]: this case is duplicates below, I think the case `SET DATA TYPE INTEGER DEFAULT NULL` is missing

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlTransactionControlParserTest.java [hunk 5]
```diff
@@ -62,14 +61,7 @@ public void testCommit() {
...
-    private static SqlNode parseStatement(String sqlStmt) {
-        return IgniteSqlParser.parse(sqlStmt, StatementParseResult.MODE).statement();
-    }
-
-    private static void expectUnparsed(SqlNode node, String expected) {
-        SqlPrettyWriter writer = new SqlPrettyWriter();
-        node.unparse(writer, 0, 0);
-
-        assertEquals(expected, writer.toString(), "Unparse does not match");
+    private SqlNode parseStatement(String sqlStmt) {
```
Comment (xtern) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlTransactionControlParserTest.java:64]: Maybe it's worth removing this method completely?

Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlTransactionControlParserTest.java]: Sure, done.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlAlterColumnDdlParserTest.java (hunk 4 above)
Comment (lowka) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/sql/SqlAlterColumnDdlParserTest.java:120]: Thanks. Fixed.

**Review by xtern — APPROVED** [review]

**Pull request merged**